core/src/main/scala/org/bitcoins/core/psbt/PSBT.scala

```diff
@@ -0,0 +1,336 @@
+import org.bitcoins.core.protocol.CompactSizeUInt
+import org.bitcoins.core.protocol.script.{
+  EmptyScriptSignature,
+  ScriptPubKey,
+  ScriptSignature,
+  ScriptWitness,
+  WitnessScriptPubKey
+}
+import org.bitcoins.core.protocol.transaction.{
+  BaseTransaction,
+  EmptyWitness,
+  Transaction,
+  WitnessTransaction
+}
+import org.bitcoins.core.script.crypto.HashType
+import scodec.bits._
+
+object PSBT {
+  final def magicBytes = hex"70736274ff"
+
+  def fromBytes(byteVector: ByteVector): PSBT = {
+    require(byteVector.take(5) == hex"70736274ff",
+            "A PSBT must start with the magic bytes 0x70736274ff")
+
+    val bytes = byteVector.drop(5)
+
+    val (bytes1: ByteVector, global: Seq[GlobalPSBTRecord], tx: Transaction) =
+      PSBTHelper.parseGlobalMap(bytes, accum = Nil, tx = null)
+    if (tx == null)
+      throw new Exception("Invalid PSBT. No global transaction")
+
+    val (bytes2, inputMaps) =
+      PSBTHelper.parseInputMaps(bytes1, tx.inputs.size, Nil)
+    val (bytes3, outputMaps) =
+      PSBTHelper.parseOutputMaps(bytes2, tx.outputs.size, Nil)
+
+    require(bytes3.isEmpty, "The PSBT should be empty now")
+
+    PSBT(GlobalPSBTMap(global.toVector),
+         inputMaps.toVector,
+         outputMaps.toVector,
+         tx)
+  }
+}
+
+case class PSBT(
+    globalMap: GlobalPSBTMap,
+    inputMaps: Vector[InputPSBTMap],
+    outputMaps: Vector[OutputPSBTMap],
+    tx: Transaction) {
+  require(inputMaps.size == tx.inputs.size)
+  require(outputMaps.size == tx.outputs.size)
+
+  val bytes: ByteVector = PSBT.magicBytes ++ globalMap.bytes ++ inputMaps
+    .foldLeft(ByteVector.empty)(_ ++ _.bytes ++ hex"ff") ++ outputMaps.foldLeft(
+    ByteVector.empty)(_ ++ _.bytes ++ hex"ff")
+
+  def toBytes: ByteVector = bytes
+}
+
+sealed trait PSBTRecord {
+
+  def key: ByteVector
+
+  def value: ByteVector
+
+  def bytes: ByteVector = {
+    val keySize = CompactSizeUInt.calc(key)
+    val valueSize = CompactSizeUInt.calc(value)
+
+    keySize.bytes ++ key ++ valueSize.bytes ++ value
+  }
+}
+
+sealed trait GlobalPSBTRecord extends PSBTRecord
+
+object GlobalPSBTRecord {
+  case class UnsignedTransaction(transaction: Transaction)
+      extends GlobalPSBTRecord {
+    require(
+      transaction.inputs.forall(_.scriptSignature == EmptyScriptSignature),
+      "All ScriptSignatures must be empty")
+    private val witnessIsEmpty = transaction match {
+      case wtx: WitnessTransaction => wtx.witness == EmptyWitness
+      case _: BaseTransaction      => true
+    }
+    require(witnessIsEmpty, "Witness must be empty")
+
+    override val key: ByteVector = hex"00"
+    override val value: ByteVector = transaction.bytes
+  }
+
+  case class XPubKey(
+      xpub: ExtPublicKey,
+      masterFingerprint: ByteVector,
+      derivationPath: BIP32Path)
+      extends GlobalPSBTRecord {
+    require(derivationPath.path.length == xpub.depth.toInt,
+            "Derivation path length does not match xpubkey depth")
```

Comment:
same comment applies for adding input that was given to the error message for easier to debugging :-)